### PR TITLE
Optimized tunable field multiplication using w-width windowed method

### PR DESCRIPTION
**Pull request by bbist**
This is a PR planned to be sent to the main BitVM repo. Please review if the interface is intuitive enough to be integrated into the main branch.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -25,19 +29,25 @@ tokio = { version = "1.37.0", features = ["full"] }
...
-bitcoin-script-stack = { git = "https://github.com/FairgateLabs/rust-bitcoin-script-stack"}
+bitcoin-script-stack = { git = "https://github.com/FairgateLabs/rust-bitcoin-script-stack" }
 rand = "0.8.5"
 rand_chacha = "0.3.1"
 once_cell = "1.19.0"
 dotenv = "0.15.0"
 aws-sdk-s3 = "1.40.0"
 regex = "1.10.5"
 blake3 = "=1.5.1"
+paste = "1.0.15"
```
Comment (storopoli) [inline Cargo.toml:40]: I think all the changes to the `Cargo.toml` that are not autoformatting (due to Taplo) we can drop.

Hence keep only this.

File: src/bigint/add.rs [hunk 2]
```diff
@@ -86,6 +87,149 @@ impl<const N_BITS: u32, const LIMB_SIZE: u32> BigIntImpl<N_BITS, LIMB_SIZE> {
...
+            { limb_double_with_carry_allow_overflow(Self::HEAD_OFFSET) }
+
+            // Take all limbs from the alt stack to the main stack
+            for _ in 0..Self::N_LIMBS - 1 {
+                OP_FROMALTSTACK
+            }
+        }
+    }
+
+    // double the item on top of the stack
```
Comment (storopoli) [inline src/bigint/add.rs:115]:
```suggestion
    /// Double the item on top of the stack
    /// 
    /// # Note
    ///
    /// This function prevents overflow of the underlying integer types during
    /// doubling operation.
```
Please as the other docstring change if necessary.

File: src/bigint/add.rs [hunk 3]
```diff
@@ -86,6 +87,149 @@ impl<const N_BITS: u32, const LIMB_SIZE: u32> BigIntImpl<N_BITS, LIMB_SIZE> {
...
+            // Take all limbs from the alt stack to the main stack
+            for _ in 0..Self::N_LIMBS - 1 {
+                OP_FROMALTSTACK
+            }
+        }
+    }
+
+    pub fn lshift_prevent_overflow(bits: u32) -> Script {
+        script! {
+            // {limb}
```
Comment (storopoli) [inline src/bigint/add.rs:141]: Is this a comment that you want to keep? Or some commented out code that you forgot to remove?

File: src/bigint/add.rs [hunk 4]
```diff
@@ -86,6 +87,149 @@ impl<const N_BITS: u32, const LIMB_SIZE: u32> BigIntImpl<N_BITS, LIMB_SIZE> {
...
+            { 1 << LIMB_SIZE } // {limb} {base}
+
+            { limb_lshift_without_carry(bits) } // {limb} {carry..} {base}
+
+            for _ in 0..Self::N_LIMBS - 2 {
+                { limb_lshift_with_carry(bits) } // {limb} {carry..} {base}
+            }
+            // // When we got {limb} {base} {carry} on the stack, we drop the base
```
Comment (storopoli) [inline src/bigint/add.rs:149]:
```suggestion
            // When we got {limb} {base} {carry} on the stack, we drop the base
```

File: src/bigint/add.rs [hunk 5]
```diff
@@ -86,6 +87,149 @@ impl<const N_BITS: u32, const LIMB_SIZE: u32> BigIntImpl<N_BITS, LIMB_SIZE> {
...
+                    }
+                }
+                for _ in 0..Self::N_LIMBS - 1 {
+                    OP_FROMALTSTACK
+                }
+            }
+        }
+    }
+
+    // does not support addition to self, stack_top=0
```
Comment (storopoli) [inline src/bigint/add.rs:191]: Maybe convert this to a docstring as well?

File: src/bigint/add.rs [hunk 6]
```diff
@@ -118,6 +262,167 @@ pub fn limb_add_nocarry(head_offset: u32) -> Script {
...
+        // {limb} {base} {carry}
+        OP_ROT // {base} {carry} {limb}
+        { NMUL(2) } // {base} {carry} {2*limb}
+        OP_ADD // {base} {2*limb + carry}
+        OP_2DUP // {base} {2*limb + carry} {base} {2*limb + carry}
+        OP_LESSTHANOREQUAL // {base} {2*limb + carry} {base<=2*limb + carry}
+        OP_TUCK // {base} {base<=2*limb+carry} {2*limb+carry} {base<=2*limb+carry}
+        OP_IF
+            2 OP_PICK OP_SUB
+        OP_ENDIF
```
Comment (storopoli) [inline src/bigint/add.rs:315]: Maybe align these the same as the others in this file?

File: src/bigint/add.rs [hunk 7]
```diff
@@ -118,6 +262,167 @@ pub fn limb_add_nocarry(head_offset: u32) -> Script {
...
+    }
+}
+
+fn limb_double_with_carry_allow_overflow(head_offset: u32) -> Script {
+    script! {
+        OP_SWAP // {carry} {limb}
+        { NMUL(2) } // {carry} {2*limb}
+        OP_ADD // {carry + 2*limb}
```
Comment (storopoli) [inline src/bigint/add.rs:323]: Maybe align these the same as the others in this file?

File: src/bigint/add.rs [hunk 8]
```diff
@@ -118,6 +262,167 @@ pub fn limb_add_nocarry(head_offset: u32) -> Script {
...
+            if i < bits { OP_ROT OP_SWAP } // {carry...} {base} {2*limb-base}
+            else { OP_TOALTSTACK OP_SWAP } // {carry...} {base} -> {2*limb-base}
+        }
+    }
+}
+
+fn limb_lshift_with_carry(bits: u32) -> Script {
+    script! {
+        // {limb} {p_carry..} {base}
+        { 1 + bits } OP_ROLL     // {p_carry..} {base} {limb}
```
Comment (storopoli) [inline src/bigint/add.rs:378]: Maybe align this the same as the others in this fn?

File: src/bigint/add.rs [hunk 9]
```diff
@@ -118,6 +262,167 @@ pub fn limb_add_nocarry(head_offset: u32) -> Script {
...
+            if i < bits { OP_ROT OP_SWAP } // {p_carry..} {carry..} {base} {2*limb-base}
+            else { OP_TOALTSTACK OP_SWAP } // {carry..} {base} -> {2*limb-base}
+        }
+    }
+}
+
+fn limb_lshift_with_carry_prevent_overflow(bits: u32, head: u32) -> Script {
+    script! {
+        // {a} {c..}
+        { bits } OP_PICK     // {a} {c..} {a}
```
Comment (storopoli) [inline src/bigint/add.rs:397]: Maybe align this the same as the others in this fn?

File: src/bigint/std.rs [hunk 10]
```diff
@@ -160,7 +160,9 @@ impl<const N_BITS: u32, const LIMB_SIZE: u32> BigIntImpl<N_BITS, LIMB_SIZE> {
     }
 
     #[inline]
-    pub fn push_zero() -> Script { push_to_stack(0, Self::N_LIMBS as usize) }
+    pub fn push_zero() -> Script {
+        push_to_stack(0, Self::N_LIMBS as usize)
+    }
```
Comment (storopoli) [inline src/bigint/std.rs:165]: Formatting change. Maybe should not be included in the PR.

File: src/bigint/std.rs [hunk 11]
```diff
@@ -237,6 +239,49 @@ impl<const N_BITS: u32, const LIMB_SIZE: u32> BigIntImpl<N_BITS, LIMB_SIZE> {
...
+
+    pub fn is_positive(depth: u32) -> Script {
+        script! {
+            { (1 + depth) * Self::N_LIMBS - 1 } OP_PICK
+            { Self::HEAD_OFFSET >> 1 }
+            OP_LESSTHAN
+        }
+    }
+
+    // resizing positive numbers; does not work for negative
```
Comment (storopoli) [inline src/bigint/std.rs:259]:
```suggestion
    /// Resize positive numbers
    ///
    /// # Note
    ///
    /// Does not work for negative numbers.
```

File: src/bn254/fq.rs [hunk 12]
```diff
@@ -14,12 +18,14 @@ impl Fp254Impl for Fq {
...
-        0x187cfd47, 0x10460b6, 0x1c72a34f, 0x2d522d0, 0x1585d978, 0x2db40c0, 0xa6e141, 0xe5c2634, 0x30644e
+        0x187cfd47, 0x10460b6, 0x1c72a34f, 0x2d522d0, 0x1585d978, 0x2db40c0, 0xa6e141, 0xe5c2634,
+        0x30644e,
     ];
 
     // inv₂₆₁ p  <=>  0x100a85dd486e7773942750342fe7cc257f6121829ae1359536782df87d1b799c77
     const MODULUS_INV_261: [u32; Self::N_LIMBS as usize] = [
-        0x1B799C77, 0x16FC3E8, 0xD654D9E, 0x30535C2, 0x257F612, 0x1A17F3E6, 0xE509D40, 0x90DCEEE, 0x100A85DD
+        0x1B799C77, 0x16FC3E8, 0xD654D9E, 0x30535C2, 0x257F612, 0x1A17F3E6, 0xE509D40, 0x90DCEEE,
+        0x100A85DD,
```
Comment (storopoli) [inline src/bn254/fq.rs:28]: Formatting change. Maybe should not be included in the PR.

File: src/bn254/fq.rs [hunk 13]
```diff
@@ -31,28 +37,295 @@ impl Fp254Impl for Fq {
...
+
+                    script! {
+                        // stack: {q} {x0} {x1} {y0} {y1}
+                        for _ in 0..2*N_LC {
+                            // range check: U < MODULUS
+                            { U::copy(0) }                                       // {q} {x0} {x1} {y0} {y1} {y1}
+                            { U::push_u32_le(&Fq::modulus_as_bigint().to_u32_digits().1) } // {q} {x0} {x1} {y0} {y1} {y1} {MODULUS}
+                            { U::lessthan(1, 0) } OP_VERIFY                      // {q} {x0} {x1} {y0} {y1}
+                            { U::toaltstack() }                                  // {q} {x0} {x1} {y0} -> {y1}
+                        }                                                              // {q} -> {x0} {x1} {y0} {y1}
```
Comment (storopoli) [inline src/bn254/fq.rs:222]: Maybe align these comments?

File: src/bn254/fq.rs [hunk 14]
```diff
@@ -31,28 +37,295 @@ impl Fp254Impl for Fq {
...
+                                { T::push_zero() }             // {q} {x0} {0} -> {x1} {y0} {y1}
+                                { T::sub(0, 1) }               // {-q} {-x0} -> {x1} {y0} {y1}
+                            }
+                            { init_table(VAR_WIDTH) }            // {-q_table} {x0_table} -> {x1} {y0} {y1}
+                        }                                                 // {-q_table} {x0_table} {x1_table} -> {y0} {y1}
+                        for _ in 0..N_LC {
+                            { U::fromaltstack() }                   // {-q_table} {x0_table} {x1_table} {y0} -> {y1}
+                            { U::resize::<{ T::N_BITS }>() } // {-q_table} {x0_table} {x1_table} {y0} -> {y1}
+                        }                                                 // {-q_table} {x0_table} {x1_table} {y0} {y1}                                                             // {q0} {q1} {x0} {x1} {y0} {y1}
+                        { T::push_zero() }                          // {-q_table} {x0_table} {x1_table} {y0} {y1} {0}
```
Comment (storopoli) [inline src/bn254/fq.rs:240]: Maybe align these comments?

File: src/bn254/fq.rs [hunk 15]
```diff
@@ -31,28 +37,295 @@ impl Fp254Impl for Fq {
...
+                        { T::copy(0) }                        // {MODULUS} {-r/r} {-r/r}
+                        { T::lessthan(0, 2) } OP_VERIFY       // {-r/r}
+
+                        // resize res back to N_BITS
+                        { T::resize::<N_BITS>() } // {r}
+                    }
+                }
+            }
+        }
+    };
```
Comment (storopoli) [inline src/bn254/fq.rs:299]: Maybe align these comments?

**Review by storopoli — COMMENTED** [review]
Some minor docstring and commenting notes.
I think that we should not upstream the Rust and TOML autoformatting changes in this PR.

Like we discussed in our call,
about the trait with the macro it is a clever design.
It is not much used in the wild but definitely can be used "in production". It reminded me `axum`'s[^1] [`all_the_tuples`](https://github.com/tokio-rs/axum/blob/1ac617a1b540e8523347f5ee889d65cad9a45ec4/axum/src/macros.rs#L49) approach.

[^1]: pretty much a good example of "in production" in Rust.

File: src/bigint/add.rs (hunk 3 above)
Comment (bbist) [reply src/bigint/add.rs]: Yeah, this is the input assumed to be on top of the stack. I changed it to, `// stack: {limb}`, to make it clear.

**Comment by bbist** [comment]
Thanks for the review @storopoli. Made most of the changes you suggested.

**Review by storopoli — APPROVED** [review]
From the Rust point of view,
ACK 04eecf105c5477ed0355be6a97fb4d633f9d6965

**Comment by bbist** [comment]
@manishbista28, can you do a quick review of the correctness of the implementation?

**Review by manishbista28 — APPROVED** [review]

**Comment by storopoli** [comment]
Closing this, feel free to reopen.
